core/plugins/org.fusesource.ide.foundation.core/src/org/fusesource/ide/foundation/core/util/JobWaiterUtil.java

```diff
@@ -51,30 +55,38 @@ private void waitJob(int decreasingCounter, IProgressMonitor monitor) {
 	}
 
 	private void joinJobs(IProgressMonitor monitor, int currentCounter) {
+		SubMonitor subMon = SubMonitor.convert(monitor, jobFamilies.size());
 		try {
 			for (Object jobFamily : jobFamilies) {
-				Job.getJobManager().join(jobFamily, monitor);
+				Job.getJobManager().join(jobFamily, subMon.split(1));
+				updateUI();
 			}
 		} catch (InterruptedException iex) {
 			// Workaround to bug
 			// https://bugs.eclipse.org/bugs/show_bug.cgi?id=335251
 			exception = iex;
-			waitJob(currentCounter, monitor);
+			waitJob(currentCounter, subMon.split(1));
 			Thread.currentThread().interrupt();			
 		} catch (OperationCanceledException e) {
 			// Workaround to bug
 			// https://bugs.eclipse.org/bugs/show_bug.cgi?id=335251
 			exception = e;
-			waitJob(currentCounter, monitor);
+			waitJob(currentCounter, subMon.split(1));
 		}
+		subMon.setWorkRemaining(0);
 	}
 
+	public static void updateUI() {
+		while (Display.getDefault().readAndDispatch()) {
```

Comment:
it seems really strange to call that. How to be sure that it doesn't stay in an infinite loop?

Reply:
I would say it is very unlikely that there is an infinite number of UI events in the event queue. The call returns true as long as there is an unprocessed UI event in the queue.

Reply:
I had more in mind that the problem can be that some UI events might trigger another UI event

Reply:
I think that we should not have need of this kind of loop in our code. There is something wrong somewhere else

Reply:
Photon still has the same problems. I am tempted to add my improvements back.